# Libraries Import

## Initializing the code by importing Libraries
      As always we first need to import all the import or those libraries that we need in this code. Bascially libraries are the premade code with alot of functions and methods to make developer job easy, mean they dont need to keep write each code all the time, they can just import it from library. Like for a sum function, if developer need to find sum of some values then he need to write a whole code for sum fucntion and then move on but by using libraries he/she doesn't need to write code for sum function as they already have it, all he/she need to do is write sum() and that operation perform. In other words Lib are made to make developer job easy.

In this code I imported two libraries **Numpy** and **Pandas**
1. Where Numpy will use to generate array 
      - Array r basically the type of variables that store more than one values inside then, like a matrix.
2. And Pandas will use to access data, manipluate it, perform operations on it like removing, inserting, replacing values etc.

In [30]:
import numpy as np
import pandas as pd

#### Reading Data
      Here I have use pandas by refering to it short name, which is pd, to read the data inside csv file name 'marketing_campaign.csv' and store it inside a variable name data.

In [81]:
data = pd.read_csv('dataset/marketing_campaign.csv')

#### Inspecting readed data
      Here I use .head() function of pandas which will display first 5 rows of dataset. I did this to inspect whether the above read code work well or not.

In [32]:
data.head()

Unnamed: 0    ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0           0  5524        1957  Graduation         Single  58138.0        0   
1           1  2174        1954  Graduation         Single  46344.0        1   
2           2  4141        1965  Graduation       Together  71613.0        0   
3           3  6182        1984  Graduation       Together  26646.0        1   
4           4  5324        1981         PhD        Married  58293.0        1   

   Teenhome Dt_Customer  Recency  ...  NumWebVisitsMonth  AcceptedCmp3  \
0         0  04-09-2012       58  ...                  7             0   
1         1  08-03-2014       38  ...                  5             0   
2         0  21-08-2013       26  ...                  4             0   
3         0  10-02-2014       26  ...                  6             0   
4         0  19-01-2014       94  ...                  5             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 30 columns]

# Filtering

#### Filtering only required Columns
      Here I filtered-out unnessary columns from the dataset and leave only required one. This let us lower the amount of total dataset and speedup the operation process, as u can see in above dataset there r around 30 columns and perfoming operation on this numbers of dataset will required alot of computation resources, and by filtering only required columns we reduce the amount of dataset, making it easier for lower systems to work with that.

In [33]:
data = data[['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency','NumStorePurchases', 'NumWebVisitsMonth']]
data.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  
0  04-09-2012       58                  4                  7  
1  08-03-2014       38                  2                  5  
2  21-08-2013       26                 10                  4  
3  10-02-2014       26                  4                  6  
4  19-01-2014       94                  6                  5

#### Shape
      This line of code will tell us the numbers of rows and columns. This help us to know amount of dataset we working with.


In [34]:
data.shape

(2240, 11)

# Grouping

#### Grouping and finding Average
      Now this line will group all the values of columns NumStorePurchases based on the row value Kidhome and then find the average of them. This is done by pandas where it done by inspecting all the row data of one Kidhome value and then finding the average of NumStorePurchases column of all these rows. Like in our case there only 3 type of values in Kidhome which r 0,1,2, pandas will first inspect all the row with value 0 and calculate the mean (average) of all the NumStorePurchases column value based on kidhome value 0 and then move to next kidhome value which is 1 and then 2.


In [35]:
data.groupby('Kidhome')['NumStorePurchases'].mean()

Kidhome
0    7.217324
1    3.863181
2    3.437500
Name: NumStorePurchases, dtype: float64

# Replacing values

      Here I converted numberic values of column 'Teenhome' to string (alphabatic value), this is done for better understanding of wheather teen present or not. This conversion is done in following way:
- 0 : 'has no teen'
- 1 & Greater than 1 : ,'has teen'
      Where 0 mean there is no teen while 1 or greater than 1 mean there r teens 
      After converting them, I store then in a new column inside the dataset and name it 'Teenhome_replaced'

In [36]:
data['Teenhome_replaced'] = data['Teenhome'].replace([0,1,2],['has no teen','has teen','has teen'])

#### Inspecting changes
      As mention above about .head() method here I used it for same purpose which is to inpect the changes

In [37]:
data.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

# Missing Values Handling

#### Finding and Droping Missing Values
      Now I will find and drop the missing values from the dataset. I will use pandas library method .isnull() to find number of missing values and then use .dropna() method to drop them.

      The Result of this code indicate that only the income column have missing values and it 24 in quantity.

In [38]:
data.isnull().sum()

ID                    0
Year_Birth            0
Education             0
Marital_Status        0
Income               24
Kidhome               0
Teenhome              0
Dt_Customer           0
Recency               0
NumStorePurchases     0
NumWebVisitsMonth     0
Teenhome_replaced     0
dtype: int64

#### Dropping Missing Values
      This line of code will drop all the missing values from dataset and store it inside another variable name 'data_withoutNa' mean data with no null value. I did this to remove the null values without altering the orignal dataset.
      I use .dropna() method of pandas to drop the missing values and tell pandas to drop any missing value by this line (how = 'any').

In [39]:
data_withoutNa = data.dropna(how = 'any')

#### Inspecting Changes
      After dropping missing I again run the same code as I did to check the missing values in orignal dataset but this time I did this with new variable that store the orignal dataset with null values. I did this to check wheather the drop command work as expected or not.
      The result of this code show that the command worked and there is no missing value now in copied dataset.

In [40]:
data_withoutNa.isnull().sum()

ID                   0
Year_Birth           0
Education            0
Marital_Status       0
Income               0
Kidhome              0
Teenhome             0
Dt_Customer          0
Recency              0
NumStorePurchases    0
NumWebVisitsMonth    0
Teenhome_replaced    0
dtype: int64

# Appending

#### Appending two Different Datasets
      Now I will demonstrate appending (merging) of two dataset into one. For this purpose I will python library pandas and it function .concat() to merge them.
      First I imported two new datasets.

In [41]:
sampleDataOne = pd.read_csv('dataset/marketing_campaign_append1.csv')
sampleDataTwo = pd.read_csv('dataset/marketing_campaign_append2.csv')

#### Inspect Data
      This line of code will tell us how many rows and columns are there in where first value show numbers of row while second one show numbers of columns

In [42]:
sampleDataOne.shape

(500, 29)

In [43]:
sampleDataTwo.shape

(500, 29)

#### Appending Data
      As I have already imported both datasets and store them in there respected variables, now I will append them and store them inside another variable name 'appendedData'. I will use pandas method concat by refereing to it short name 'pd' and give both sample data variables as argument ( values).

In [44]:
appendedData = pd.concat( [sampleDataOne, sampleDataTwo])

#### Inspecting Changes
      This line of code demonstrate the changes being made and display number of rows and columns.
      As u see above that number of rows in each sample datasets r 500 and after appending them it become 1000, this show that the above command was successful.

In [45]:
appendedData.shape

(1000, 29)

# Concatenation
      Concatenation and append sometime look similar but they r not. The main difference between them is that:
            Concatenation combine multiple dataframes with another while appending can only add another dataframe at the end of first one.
            While both use concat method of python because append method is depricated.
      
      Now I will demonstrate the Concatination of datasets.
      First I add imported all the data of from orignal dataset to both sample variables.

In [46]:
sampleDataOne = data
sampleDataOne

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  \
0            0  04-09-2012       58                  4                  7   
1            1  08-03-2014       38                  2                  5   
2            0  21-08-2013       26                 10                  4   
3            0  10-02-2014       26                  4                  6   
4            0  19-01-2014       94                  6                  5   
...        ...         ...      ...                ...                ...   
2235         1  13-06-2013       46                  4                  5   
2236         1  10-06-2014       56                  5                  7   
2237         0  25-01-2014       91                 13                  6   
2238         1  24-01-2014        8                 10                  3   
2239         1  15-10-2012       40                  4                  7   

     Teenhome_replaced  
0          has no teen  
1             has teen  
2          has no teen  
3          has no teen  
4          has no teen  
...                ...  
2235          has teen  
2236          has teen  
2237       has no teen  
2238          has teen  
2239          has teen  

[2240 rows x 12 columns]

In [47]:
sampleDataTwo = data
sampleDataTwo.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

#### Filtering
      Now I have dropped some columns from both samples. This is done to demonstrate concatenations.
      First I will drop last 6 from first sample and then drop first 6 from second sample

In [48]:
sampleDataOne = sampleDataOne.drop(['Teenhome', 'Dt_Customer', 'Recency', 'NumStorePurchases', 'NumWebVisitsMonth','Teenhome_replaced'], axis=1)
sampleDataOne.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome
0  5524        1957  Graduation         Single  58138.0        0
1  2174        1954  Graduation         Single  46344.0        1
2  4141        1965  Graduation       Together  71613.0        0
3  6182        1984  Graduation       Together  26646.0        1
4  5324        1981         PhD        Married  58293.0        1

In [49]:
sampleDataTwo = sampleDataTwo.drop(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income','Kidhome'], axis=1)

In [50]:
sampleDataTwo.head()

Teenhome Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  \
0         0  04-09-2012       58                  4                  7   
1         1  08-03-2014       38                  2                  5   
2         0  21-08-2013       26                 10                  4   
3         0  10-02-2014       26                  4                  6   
4         0  19-01-2014       94                  6                  5   

  Teenhome_replaced  
0       has no teen  
1          has teen  
2       has no teen  
3       has no teen  
4       has no teen

#### Concatenation
      Now I have concatenate both samples with 6 6 columns into one single variabel. I use pandas .concat() method for this purpose.
      In this code I also mention concatenation direction I mean how the concatenation should happen i.e row wise or columns wise. In this code the line 'axis = 1' indicating that the concatenation should be happen column wise.

In [51]:
appendedData = pd.concat([sampleDataOne,sampleDataTwo],axis=1)

#### Output Inspection
      And at last I use .head() method to inspect the changes.
      In below output there r total 12 columns which indication successfull running of concat commad.

In [52]:
appendedData.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

# Merging

#### Difference Between Merging, appending, concatenation
      Sometime all these three terms sound similar but they r not. each one do it work in different way, the main difference is following:
| Name     | Difference |
|----------|----------|
| Appending    | It add second dataset at the end of first one     |
| Concatenation    | It combine multiple frames rows, and columns wise     |
| Merging     | It combine columns values based on specifi column value |


Now I will demonstrate merging of two datasets
for that purpose I will first drop some columns from one dataset and some from second one.
Also I shuffle whole dataset for legit demonstration

In [62]:
sampleDataOne = data.copy()

In [54]:
data.dtypes

ID                     int64
Year_Birth             int64
Education             object
Marital_Status        object
Income               float64
Kidhome                int64
Teenhome               int64
Dt_Customer           object
Recency                int64
NumStorePurchases      int64
NumWebVisitsMonth      int64
Teenhome_replaced     object
dtype: object

In [63]:
sampleDataOne = (sampleDataOne.drop([
      'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome',
      'Dt_Customer', 'Recency', 'NumStorePurchases',
      'NumWebVisitsMonth', 'Teenhome_replaced'
], axis=1))

In [56]:
sampleDataOne.head()

ID Year_Birth
0  5524       1957
1  2174       1954
2  4141       1965
3  6182       1984
4  5324       1981

In [64]:
sampleDataTwo = data.copy()
sampleDataTwo.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

In [66]:
sampleDataTwo = sampleDataTwo.drop([
    'Year_Birth', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency',
    'NumStorePurchases', 'NumWebVisitsMonth', 'Teenhome_replaced'
], axis=1)

sampleDataTwo.head()

ID   Education Marital_Status   Income
0  5524  Graduation         Single  58138.0
1  2174  Graduation         Single  46344.0
2  4141  Graduation       Together  71613.0
3  6182  Graduation       Together  26646.0
4  5324         PhD        Married  58293.0

Now I will merge both samples

In [67]:
Merge_data = pd.merge(sampleDataOne, sampleDataTwo, on = 'ID')

In [68]:
Merge_data.head()

ID  Year_Birth   Education Marital_Status   Income
0  5524        1957  Graduation         Single  58138.0
1  2174        1954  Graduation         Single  46344.0
2  4141        1965  Graduation       Together  71613.0
3  6182        1984  Graduation       Together  26646.0
4  5324        1981         PhD        Married  58293.0

# Sorting

Now I will demonstrate sorting of a dataset based on a value of specified columns. For that purpose I will use pandas method 'sort_value()'. This method help in quick sorting by providing sorting order, column for sorting.
For that reason I will first create a new variable for sorting and assign the orignal dataset to it

In [69]:
sorted_dataset = data.copy()
sorted_dataset.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

#### Filtering
      Now I will filter out only the required columns while unnessary columns 

In [70]:
sorted_dataset = sorted_dataset[['ID', 'Year_Birth', 'Education', 'Marital_Status',
                  'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 
                  'Recency','NumStorePurchases', 'NumWebVisitsMonth']]
sorted_dataset.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  
0  04-09-2012       58                  4                  7  
1  08-03-2014       38                  2                  5  
2  21-08-2013       26                 10                  4  
3  10-02-2014       26                  4                  6  
4  19-01-2014       94                  6                  5

#### Sorting 
      Now I will sort dataset based on columns 'NumStorePurchases' and by using pandas method '.sort_values()'

In [73]:
sorted_dataset = sorted_dataset.sort_values('NumStorePurchases', ascending=True)
sorted_dataset

ID  Year_Birth   Education Marital_Status    Income  Kidhome  \
1846   9931        1963         PhD        Married    4023.0        1   
1653   4931        1977  Graduation       Together  157146.0        0   
1524  11110        1973  Graduation         Single    3502.0        1   
1042  10749        1991  Graduation         Single    8028.0        0   
27     5255        1986  Graduation         Single       NaN        1   
...     ...         ...         ...            ...       ...      ...   
135    1685        1967         PhD       Together   62981.0        0   
1325   9212        1962      Master        Married   72828.0        0   
485    2426        1964         PhD       Divorced   78825.0        0   
1485   9988        1976      Master         Single   70379.0        0   
1801   6977        1974  Graduation       Together   75702.0        0   

      Teenhome Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  
1846         1  23-06-2014       29                  0                 19  
1653         0  29-04-2013       13                  0                  1  
1524         0  13-04-2013       56                  0                 14  
1042         0  18-09-2012       62                  0                 19  
27           0  20-02-2013       19                  0                  1  
...        ...         ...      ...                ...                ...  
135          0  17-03-2013       21                 13                  3  
1325         1  14-02-2013       17                 13                  7  
485          0  14-09-2012       35                 13                  3  
1485         1  03-03-2013       84                 13                  4  
1801         1  14-10-2012       87                 13                  6  

[2240 rows x 11 columns]

# Categorise 
      Now I will show the dataset categorizing based on the values.
      In the following code I will assign a specif value based on previous value.
      And this will be done using pandas method '.cut()' 

In [74]:
sampleDataOne = data.copy()
sampleDataOne.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth Teenhome_replaced  
0  04-09-2012       58                  4                  7       has no teen  
1  08-03-2014       38                  2                  5          has teen  
2  21-08-2013       26                 10                  4       has no teen  
3  10-02-2014       26                  4                  6       has no teen  
4  19-01-2014       94                  6                  5       has no teen

In [75]:
sampleDataOne = sampleDataOne[['ID', 'Year_Birth', 'Education', 'Marital_Status',
                  'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 
                  'Recency','NumStorePurchases', 'NumWebVisitsMonth']]
sampleDataOne.head()

ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  
0  04-09-2012       58                  4                  7  
1  08-03-2014       38                  2                  5  
2  21-08-2013       26                 10                  4  
3  10-02-2014       26                  4                  6  
4  19-01-2014       94                  6                  5

#### Categorising 
      In the following code I first copied original data to categorized data name varible for categrization demonstration and then i use panda (by refereing to it short name pd) mehtod cut that handling this categorization related things and give it following arragument(values):
            x = sampleDataOne['NumStorePurchases'] 
                  To categorize value based on values of this columns
            bins=[0,4,8,13]
                  These r the values range upon which categorization will be done 
            labels=['low','Medium','High']
                  these r the labels that will be give to the value based on the range given in bins list
            
            let say a value in column NumStorePurchases is 6 and 6 comes before 8 then the value in that row under column bin will be medium.

In [78]:
categorizedData = data.copy()
categorizedData ['bin'] = pd.cut(x = sampleDataOne['NumStorePurchases'],bins=[0,4,8,13],labels=['low','Medium','High'])
categorizedData = categorizedData.sort_values('bin', ascending=True)
categorizedData.head()


ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1244   1291        1978         PhD       Together  27683.0        1   
1241   1921        1967         PhD        Married  59062.0        2   
1240  11086        1984       Basic         Single  25707.0        1   
1238   6918        1989  Graduation        Married  28691.0        1   

      Teenhome Dt_Customer  Recency  NumStorePurchases  NumWebVisitsMonth  bin  
0            0  04-09-2012       58                  4                  7  low  
1244         0  04-08-2012       90                  4                  8  low  
1241         1  03-10-2013       74                  3                  4  low  
1240         0  18-01-2014       18                  3                  7  low  
1238         0  04-07-2013       56                  3                  8  low

# Duplicate Removal
      Now I will show how to remove duplicated data from dataset.
      In the following code I will pandas methods .duplicate() to check for avalible duplicates and .drop_duplicate() to remove those rows.For this demonstration I will use another dataset which is about movies including there name, genere, stars, year of release and one line story. I did this because the previous dataset doesn't contain any duplicate


In [114]:
moviesData = pd.read_csv('dataset/movies.csv')

The output of this code show that there r around 431 duplicated rows. Now I will remove that using drop_duplicate() method

In [89]:
moviesData.duplicated().sum()

431

In [92]:
moviesData.shape

(9999, 9)

In [93]:
moviesData_no_duplicate = moviesData.drop_duplicates()

After removing again I use same function as to find the duplicated rows but this time it say 0 which mean there no more duplicated rows left after drop command

In [94]:
moviesData_no_duplicate.duplicated().sum()

0

# Row and Column Removal
      In the below code I will show how we can remove row or column from the dataset.
      for that purpose I will use pandas drop function as before but this time with some extra arrgments (values) like name of row of column to be drop etc.


In [95]:
moviesData.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                            ONE-LINE  \
0  \nA woman with a mysterious illness is forced ...   
1  \nThe war for Eternia begins again in what may...   
2  \nSheriff Deputy Rick Grimes wakes up from a c...   
3  \nAn animated series that follows the exploits...   
4  \nA prequel, set before the events of Army of ...   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN

In the following code, I removed a column name 'ONE-LINE' by giving command 
      moviesData = moviesData.drop(labels='ONE-LINE',axis=1)
where labels='ONE-LINE' telling the name to be remove, while
axis=1 tell the direction of dataset value where axis 1 mean column and 0 mean row.

In [99]:
moviesData = moviesData.drop(labels='ONE-LINE',axis=1)
moviesData.head()

MOVIES         YEAR  \
0                        Blood Red Sky       (2021)   
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   

                                        GENRE  RATING  \
0      \nAction, Horror, Thriller                 6.1   
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   

                                               STARS    VOTES  RunTime Gross  
0  \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   NaN  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN

Now I will Remova a row using same method but with different values.
The result of this command show that first row which was:
| Title           | Year | Genre                      | Rating | Description                                                                                           | Director         | Stars                                                                  | Votes  | Runtime |
|-----------------|------|----------------------------|--------|-------------------------------------------------------------------------------------------------------|------------------|------------------------------------------------------------------------|--------|---------|
| Blood Red Sky   | 2021 | Action, Horror, Thriller   | 6.1    | A woman with a mysterious illness is forced into action when a group of terrorists hijack a flight.   | Peter Thorwarth  | Peri Baumeister, Carl Anton Koch, Alexander Scheer, Kais Setti        | 21,062 | 121     |


but after this command run that row disappeared which mean that command run successfully

In [100]:
moviesData = moviesData.drop(labels=0, axis=0)
moviesData.head()

MOVIES         YEAR  \
1  Masters of the Universe: Revelation     (2021– )   
2                     The Walking Dead  (2010–2022)   
3                       Rick and Morty     (2013– )   
4                      Army of Thieves       (2021)   
5                          Outer Banks     (2020– )   

                                        GENRE  RATING  \
1  \nAnimation, Action, Adventure                 5.0   
2       \nDrama, Horror, Thriller                 8.2   
3  \nAnimation, Adventure, Comedy                 9.2   
4         \nAction, Crime, Horror                 NaN   
5          \nAction, Crime, Drama                 7.6   

                                               STARS    VOTES  RunTime Gross  
1  \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   NaN  
2  \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   NaN  
3  \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   NaN  
4  \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   NaN  
5  \n            \n    Stars:\nChase Stokes, \nMa...   25,858     50.0   NaN

# Format Changing
      In this code I will change the datatype of a column inside the dataset. For that purpose I will use movies dataset, and the method I will use is of pandas of name 'astype' and 'dtype' where astype change the type while dtype tell the current datatype
      

In [102]:
moviesData.dtypes

MOVIES      object
YEAR        object
GENRE       object
RATING     float64
STARS       object
VOTES       object
RunTime    float64
Gross       object
dtype: object

I will change the datatype of column RunTime which is float to int

In [116]:
moviesData['RunTime'] = moviesData['RunTime'].fillna('0')

In [117]:
moviesData['RunTime'] = moviesData['RunTime'].astype(int)

In [118]:
moviesData.dtypes

MOVIES       object
YEAR         object
GENRE        object
RATING      float64
ONE-LINE     object
STARS        object
VOTES        object
RunTime       int32
Gross        object
dtype: object